File: frontend/src/types/reward.ts
/**
 * Reward domain types.
 *
 * Mirror the backend contract for the rewards module exactly
 * (backend/src/modules/rewards/*, docs/engineering/05_API.md). Keep in sync
 * when the backend contract changes.
 */

/** Reasons a reward transaction can be issued (Prisma `RewardReason`). */
export const REWARD_REASONS = [
  'RECYCLING',
  'BONUS',
  'CAMPAIGN',
  'ADJUSTMENT',
  'REDEMPTION',
] as const;

export type RewardReason = (typeof REWARD_REASONS)[number];

/**
 * Balance snapshot returned by GET /rewards/balance (`RewardBalance`).
 * `greenCoins` is the spendable balance; the totals are cumulative lifetime
 * metrics computed from the caller's reward history.
 */
export interface RewardBalance {
  greenCoins: number;
  totalRewards: number;
  totalCO2Saved: number;
  totalEnergySaved: number;
  totalLandfillDiverted: number;
}

/**
 * A single reward transaction with its related submission context, as returned
 * by GET /rewards/history (`RewardTransactionWithSubmission`). `createdAt`
 * fields are ISO strings.
 */
export interface RewardHistoryItem {
  id: string;
  userId: string;
  submissionId: string;
  points: number;
  reason: RewardReason;
  createdAt: string;
  submission: {
    id: string;
    category: string;
    status: string;
    estimatedWeight: number;
    createdAt: string;
  };
}

File: mobile/consumer_app/lib/features/rewards/models/reward.dart
/// Mirrors `RewardBalance` (`backend/src/modules/rewards/reward.service.ts`).
class RewardBalance {
  const RewardBalance({
    required this.greenCoins,
    required this.totalRewards,
    required this.totalCO2Saved,
    required this.totalEnergySaved,
    required this.totalLandfillDiverted,
  });

  final int greenCoins;
  final int totalRewards;
  final double totalCO2Saved;
  final double totalEnergySaved;
  final double totalLandfillDiverted;

  factory RewardBalance.fromJson(Map<String, dynamic> json) {
    return RewardBalance(
      greenCoins: json['greenCoins'] as int,
      totalRewards: json['totalRewards'] as int,
      totalCO2Saved: (json['totalCO2Saved'] as num).toDouble(),
      totalEnergySaved: (json['totalEnergySaved'] as num).toDouble(),
      totalLandfillDiverted: (json['totalLandfillDiverted'] as num).toDouble(),
    );
  }
}

/// Mirrors `RewardTransactionWithSubmission`
/// (`backend/src/modules/rewards/reward.repository.ts`).
class RewardTransaction {
  const RewardTransaction({
    required this.id,
    required this.submissionId,
    required this.points,
    required this.reason,
    required this.createdAt,
    required this.submissionCategory,
    required this.submissionStatus,
  });

  final String id;
  final String submissionId;
  final int points;

  /// One of `RECYCLING | BONUS | CAMPAIGN | ADJUSTMENT | REDEMPTION`
  /// (`RewardReason` enum, `backend/prisma/schema.prisma`).
  final String reason;
  final DateTime createdAt;
  final String submissionCategory;
  final String submissionStatus;

  factory RewardTransaction.fromJson(Map<String, dynamic> json) {
    final submission = json['submission'] as Map<String, dynamic>;
    return RewardTransaction(
      id: json['id'] as String,
      submissionId: json['submissionId'] as String,
      points: json['points'] as int,
      reason: json['reason'] as String,
      createdAt: DateTime.parse(json['createdAt'] as String),
      submissionCategory: submission['category'] as String,
      submissionStatus: submission['status'] as String,
    );
  }
}
